File: App/Resources/AlgorithmDetails/blockinsertionsort/blockinsertionsort.rb
def multi_swap(arr, a, b, count)
  (0...count).each do |i|
    arr[a + i], arr[b + i] = arr[b + i], arr[a + i]
  end
end

def rotate(arr, pos, len_a, len_b)
  while len_a != 0 && len_b != 0
    if len_a <= len_b
      multi_swap(arr, pos, pos + len_a, len_a)
      pos += len_a
      len_b -= len_a
    else
      multi_swap(arr, pos + (len_a - len_b), pos + len_a, len_b)
      len_a -= len_b
    end
  end
end

def bin_search(arr, pos, length, key_pos, is_left)
  left = 0
  right = length
  while left < right
    mid = left + (right - left) / 2
    cond = is_left ? arr[pos + mid] < arr[key_pos] : arr[pos + mid] <= arr[key_pos]
    if cond
      left = mid + 1
    else
      right = mid
    end
  end
  left
end

def merge_without_buffer(arr, pos, len1, len2)
  return if len1 == 0 || len2 == 0

  if len1 == 1
    loc = bin_search(arr, pos + 1, len2, pos, true)
    rotate(arr, pos, 1, loc)
    return
  end
  if len2 == 1
    loc = bin_search(arr, pos, len1, pos + len1, false)
    rotate(arr, pos + loc, len1 - loc, 1)
    return
  end
  mid1 = len1 / 2
  loc = bin_search(arr, pos + len1, len2, pos + mid1, true)
  rotate(arr, pos + mid1, len1 - mid1, loc)
  merge_without_buffer(arr, pos, mid1, loc)
  merge_without_buffer(arr, pos + mid1 + loc, len1 - mid1, len2 - loc)
end

def find_run(arr, a, b)
  i = a + 1
  return i if i == b

  if arr[i - 1] > arr[i]
    i += 1
    i += 1 while i < b && arr[i - 1] > arr[i]
    lo = a
    hi = i - 1
    while lo < hi
      arr[lo], arr[hi] = arr[hi], arr[lo]
      lo += 1
      hi -= 1
    end
  else
    i += 1
    i += 1 while i < b && arr[i - 1] <= arr[i]
  end
  i
end

def insert1(arr, a, l)
  tmp = arr[l]
  l -= 1
  while l >= a && arr[l] > tmp
    arr[l + 1] = arr[l]
    l -= 1
  end
  arr[l + 1] = tmp
end

def insert2(arr, a, l, r)
  tmp_l = arr[l]
  tmp_r = arr[r]
  l -= 1
  while l >= a && arr[l] > tmp_r
    arr[l + 2] = arr[l]
    l -= 1
  end
  arr[l + 2] = tmp_r
  while l >= a && arr[l] > tmp_l
    arr[l + 1] = arr[l]
    l -= 1
  end
  arr[l + 1] = tmp_l
end

def sort(arr)
  n = arr.length
  i = find_run(arr, 0, n)
  while i < n
    j = find_run(arr, i, n)
    len = j - i
    if len == 1
      insert1(arr, 0, i)
    elsif len == 2
      insert2(arr, 0, i, i + 1)
    else
      merge_without_buffer(arr, 0, i, len)
    end
    i = j
  end
end

array = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56]
sort(array)
p array

File: App/Resources/AlgorithmDetails/blockinsertionsort/blockinsertionsort.swift
func multiSwap(_ arr: inout [Int], _ a: Int, _ b: Int, _ count: Int) {
    for i in 0 ..< count {
        arr.swapAt(a + i, b + i)
    }
}

func rotate(_ arr: inout [Int], _ pos: Int, _ lenA: Int, _ lenB: Int) {
    var p = pos
    var la = lenA
    var lb = lenB
    while la != 0, lb != 0 {
        if la <= lb {
            multiSwap(&arr, p, p + la, la)
            p += la
            lb -= la
        } else {
            multiSwap(&arr, p + (la - lb), p + la, lb)
            la -= lb
        }
    }
}

func binSearch(_ arr: [Int], _ pos: Int, _ len: Int, _ keyPos: Int, _ isLeft: Bool) -> Int {
    var left = 0
    var right = len
    while left < right {
        let mid = left + (right - left) / 2
        let cond = isLeft ? arr[pos + mid] < arr[keyPos] : arr[pos + mid] <= arr[keyPos]
        if cond {
            left = mid + 1
        } else {
            right = mid
        }
    }
    return left
}

func mergeWithoutBuffer(_ arr: inout [Int], _ pos: Int, _ len1: Int, _ len2: Int) {
    if len1 == 0 || len2 == 0 {
        return
    }
    if len1 == 1 {
        let loc = binSearch(arr, pos + 1, len2, pos, true)
        rotate(&arr, pos, 1, loc)
        return
    }
    if len2 == 1 {
        let loc = binSearch(arr, pos, len1, pos + len1, false)
        rotate(&arr, pos + loc, len1 - loc, 1)
        return
    }
    let mid1 = len1 / 2
    let loc = binSearch(arr, pos + len1, len2, pos + mid1, true)
    rotate(&arr, pos + mid1, len1 - mid1, loc)
    mergeWithoutBuffer(&arr, pos, mid1, loc)
    mergeWithoutBuffer(&arr, pos + mid1 + loc, len1 - mid1, len2 - loc)
}

func findRun(_ arr: inout [Int], _ a: Int, _ b: Int) -> Int {
    var i = a + 1
    if i == b {
        return i
    }
    if arr[i - 1] > arr[i] {
        i += 1
        while i < b && arr[i - 1] > arr[i] {
            i += 1
        }
        var lo = a
        var hi = i - 1
        while lo < hi {
            arr.swapAt(lo, hi)
            lo += 1
            hi -= 1
        }
    } else {
        i += 1
        while i < b && arr[i - 1] <= arr[i] {
            i += 1
        }
    }
    return i
}

func insert1(_ arr: inout [Int], _ a: Int, _ l: Int) {
    let tmp = arr[l]
    var i = l - 1
    while i >= a, arr[i] > tmp {
        arr[i + 1] = arr[i]
        i -= 1
    }
    arr[i + 1] = tmp
}

func insert2(_ arr: inout [Int], _ a: Int, _ l: Int, _ r: Int) {
    let tmpL = arr[l]
    let tmpR = arr[r]
    var i = l - 1
    while i >= a, arr[i] > tmpR {
        arr[i + 2] = arr[i]
        i -= 1
    }
    arr[i + 2] = tmpR
    while i >= a, arr[i] > tmpL {
        arr[i + 1] = arr[i]
        i -= 1
    }
    arr[i + 1] = tmpL
}

func sort(_ arr: inout [Int]) {
    let n = arr.count
    var i = findRun(&arr, 0, n)
    while i < n {
        let j = findRun(&arr, i, n)
        let len = j - i
        if len == 1 {
            insert1(&arr, 0, i)
        } else if len == 2 {
            insert2(&arr, 0, i, i + 1)
        } else {
            mergeWithoutBuffer(&arr, 0, i, len)
        }
        i = j
    }
}

var array: [Int] = [
    0, 39, 21, 62, 91, 77, 14, 23,
    90, 69, 51, 81, 68, 83, 32, 56,
]
sort(&array)
print(array)
